library/classes/thumbnail/ThumbnailGenerator.php

```diff
@@ -0,0 +1,184 @@
+    }
+
+    /**
+     * @return array
+     */
+    public static function get_types_support()
+    {
+        foreach(self::$types_support as $value){
+            $types_support[] = "'$value'";
+        }
+        return $types_support;
+    }
+
+    static function count_not_generated()
+    {
+        $sql = "SELECT COUNT(*) AS c FROM documents
+        WHERE mimetype IN (" . implode(',', self::get_types_support()) . ") AND thumb_url IS NULL";
+
+        $results = sqlStatement($sql);
+        $row = sqlFetchArray($results);
+
+        return $row['c'];
+    }
+
+    /**
+     * Generating all files with match format that still isn't generated
+     * @return array $result (count failed/success)
+     */
+    public function generate_all()
+    {
+
+        $feedback = array('sum_success' => 0, 'sum_failed' => 0, 'success' => array(), 'failed' => array());
+
+        $sql = "SELECT id, url, couch_docid, storagemethod FROM documents
+        WHERE mimetype IN (" . implode(',', self::get_types_support()) . ") AND thumb_url IS NULL";
+
+        $results = sqlStatement($sql);
+        while($row = sqlFetchArray($results)) {
+
+            switch((int)$row['storagemethod']) {
+                //for hard disk store
+                case 0:
+                    $new_file = $this->generate_HD_file($row['url']);
+                    break;
+                //for CouchDB store
+                case 1:
+                    $new_file =  $this->generate_couch_file($row['couch_docid'], $row['url']);
+                    break;
+                default:
+                    $this->error_log($row['url']);continue;
+            }
+            // Write error to log if failed
+            if (!$new_file) {
+                $this->error_log($row['url']);
+                $feedback['sum_failed'] ++;
+                $feedback['failed'][] = $row['url'];
+                continue;
+            }
+
+            $sql = "UPDATE documents SET thumb_url = ? WHERE id = ?";
+            $update = sqlStatement($sql, array($new_file, $row['id']));
+            if($update) {
+               $feedback['sum_success'] ++;
+               $feedback['success'][] = $row['url'];
+            }
+        }
+
+        return $feedback;
+    }
+
+    /**
+     * Generate new file and store it in hard disk
+     * @param $url
+     * @return bool|string
+     */
+    private function generate_HD_file($url)
+    {
+        //remove 'file://'
+        $url = substr($url, 7);
+        $path_parts = pathinfo($url);
+
+        if(!is_file($url) || empty($path_parts['extension'])) {
+            return false;
+        }
+
+        $resource = $this->thumb_obj->create_thumbnail($url);
+        if(!$resource) {
+            return false;
+        }
+
+        $thumb_name = $this->get_thumb_name($path_parts['basename']);
+        $full_path = $path_parts['dirname'] . '/' . $thumb_name;
+        $new_thumb_file = $this->thumb_obj->image_to_file($resource, $full_path);
```

Comment:
It doesn't look like you are correctly figuring out the path. See the algorithm in the documents class for displaying and saving files to ensure done correctly (note it's not simply a matter of copying the url or else it would break if changed the path of openemr).